### PR TITLE
CC-14636 UnzerApi module enabler

### PR DESCRIPTION
**Pull request by kkicheglovspryker**
Ticket: https://spryker.atlassian.net/browse/CC-14636

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# UnzerApi Module
+[![CI](https://github.com/spryker-eco/unzer-api/actions/workflows/ci.yml/badge.svg)](https://github.com/spryker-eco/unzer-api/actions/workflows/ci.yml)
+[![Latest Stable Version](https://poser.pugx.org/spryker-eco/unzer-api/v/stable.svg)](https://packagist.org/packages/spryker-eco/unzer-api)
+[![License](https://img.shields.io/github/license/spryker-eco/unzer-api.svg?b=master)](https://github.com/spryker-eco/unzer-api)
+[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/spryker-eco/unzer-api/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/spryker-eco/unzer-api/?branch=master)
+[![Build Status](https://scrutinizer-ci.com/g/spryker-eco/unzer-api/badges/build.png?b=master)](https://scrutinizer-ci.com/g/spryker-eco/unzer-api/build-status/master)
+[![Minimum PHP Version](https://img.shields.io/badge/php-%3E%3D%207.3-8892BF.svg)](https://php.net/)
+## Installation
```
Comment (AsonUnique) [inline README.md:11]: Please, add small description of the module here.

File: src/SprykerEco/Zed/UnzerApi/Business/Api/Client/UnzerApiClient.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
+<?php declare(strict_types = 1);
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Business/Api/Client/UnzerApiClient.php]:
```suggestion
<?php
```

File: src/SprykerEco/Zed/UnzerApi/Business/Api/Logger/UnzerApiLogger.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+use SprykerEco\Zed\UnzerApi\Persistence\UnzerApiEntityManagerInterface;
+
+class UnzerApiLogger implements UnzerApiLoggerInterface
+{
+    use TransactionTrait;
+
+    /**
+     * @var \SprykerEco\Zed\UnzerApi\Persistence\UnzerApiEntityManagerInterface
+     */
+    protected $entityManager;
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Business/Api/Logger/UnzerApiLogger.php]:
```suggestion
    protected $unzerApiEntityManager;
```

File: src/SprykerEco/Zed/UnzerApi/Business/Api/Logger/UnzerApiLogger.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        string $url
+    ): void {
+        $paymentUnzerApiLog = $this->createPaymentUnzerApiLogTransfer(
+            $unzerApiRequestTransfer,
+            $unzerApiResponseTransfer,
+            $requestType,
+            $url
+        );
+
+        $this->getTransactionHandler()->handleTransaction(function () use ($paymentUnzerApiLog) {
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Business/Api/Logger/UnzerApiLogger.php]: I don't think that you need a transaction here.

File: src/SprykerEco/Zed/UnzerApi/Dependency/External/Guzzle/UnzerApiToGuzzleHttpClientAdapterInterface.php [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Zed\UnzerApi\Dependency\External\Guzzle;
+
+use SprykerEco\Zed\UnzerApi\Dependency\External\Guzzle\Response\UnzerApiGuzzleResponseInterface;
+
+interface UnzerApiToGuzzleHttpClientAdapterInterface
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Dependency/External/Guzzle/UnzerApiToGuzzleHttpClientAdapterInterface.php]:
```suggestion
interface UnzerApiToHttpClientAdapterInterface
```

According to RFC https://spryker.atlassian.net/wiki/spaces/RFC/pages/18350081/ACCEPTED+RFC+Adapters+to+vendor+classes

File: src/SprykerEco/Zed/UnzerApi/Dependency/External/Guzzle/Exception/UnzerApiGuzzleRequestException.php [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Zed\UnzerApi\Dependency\External\Guzzle\Exception;
+
+use Exception;
+use SprykerEco\Zed\UnzerApi\Dependency\External\Guzzle\Response\UnzerApiGuzzleResponseInterface;
+use Throwable;
+
+class UnzerApiGuzzleRequestException extends Exception
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Dependency/External/Guzzle/Exception/UnzerApiGuzzleRequestException.php]: I'd rather not mention the third-party library names in our class names.

File: src/SprykerEco/Zed/UnzerApi/Dependency/External/Guzzle/Response/UnzerApiGuzzleResponseInterface.php [hunk 7]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Zed\UnzerApi\Dependency\External\Guzzle\Response;
+
+interface UnzerApiGuzzleResponseInterface
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Dependency/External/Guzzle/Response/UnzerApiGuzzleResponseInterface.php]:
```suggestion
interface UnzerApiResponseInterface
```

File: src/SprykerEco/Zed/UnzerApi/Dependency/External/Guzzle/Response/UnzerApiGuzzleResponseInterface.php [hunk 8]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Zed\UnzerApi\Dependency\External\Guzzle\Response;
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Dependency/External/Guzzle/Response/UnzerApiGuzzleResponseInterface.php]: I don't think that you need to create folders `Guzzle` and `Response`.

File: src/SprykerEco/Zed/UnzerApi/Dependency/External/Guzzle/UnzerApiToToGuzzleHttpClientAdapter.php [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+        return $this->createUnzerApiGuzzleResponse($response);
+    }
+
+    /**
+     * @param \Psr\Http\Message\ResponseInterface|null $response
+     *
+     * @return \SprykerEco\Zed\UnzerApi\Dependency\External\Guzzle\Response\UnzerApiGuzzleResponse
+     */
+    protected function createUnzerApiGuzzleResponse(?ResponseInterface $response): UnzerApiGuzzleResponse
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Dependency/External/Guzzle/UnzerApiToToGuzzleHttpClientAdapter.php]:
```suggestion
    protected function createUnzerApiGuzzleResponse(?ResponseInterface $response): UnzerApiResponseInterface
```

File: src/SprykerEco/Zed/UnzerApi/Persistence/UnzerApiQueryContainer.php [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+ */
+
+namespace SprykerEco\Zed\UnzerApi\Persistence;
+
+use Spryker\Zed\Kernel\Persistence\AbstractQueryContainer;
+
+/**
+ * @method \SprykerEco\Zed\UnzerApi\Persistence\UnzerApiPersistenceFactory getFactory()
+ */
+class UnzerApiQueryContainer extends AbstractQueryContainer implements UnzerApiQueryContainerInterface
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Persistence/UnzerApiQueryContainer.php]: Please remove this class and its interface

File: tooling.yml [hunk 11]
```diff
@@ -0,0 +1,2 @@
+architecture-sniffer:
```
Comment (AsonUnique) [inline tooling.yml:1]: Please, add code-sniffer here too

File: README.md (hunk 1 above)
Comment (kkicheglovspryker) [reply README.md:11]: done

File: tooling.yml (hunk 11 above)
Comment (kkicheglovspryker) [reply tooling.yml:1]: done

File: src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+use Spryker\Service\UtilEncoding\UtilEncodingServiceInterface;
+use SprykerEco\Zed\UnzerApi\Business\Api\Response\Mapper\UnzerApiResponseMapperInterface;
+use SprykerEco\Zed\UnzerApi\Dependency\External\Guzzle\Response\UnzerApiResponseInterface;
+
+class UnzerApiResponseConverter implements UnzerApiResponseConverterInterface
+{
+    /**
+     * @var \Spryker\Service\UtilEncoding\UtilEncodingServiceInterface
+     */
+    protected $encodingService;
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php]:
```suggestion
    protected $utilEncodingService;
```

File: src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+    /**
+     * @var \SprykerEco\Zed\UnzerApi\Business\Api\Response\Mapper\UnzerApiResponseMapperInterface
+     */
+    protected $responseMapper;
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php]:
```suggestion
    protected $unzerApiResponseMapper;
```

File: src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    /**
+     * @param \SprykerEco\Zed\UnzerApi\Dependency\External\Guzzle\Response\UnzerApiGuzzleResponseInterface $response
+     * @param bool $isSuccess
+     *
+     * @return \Generated\Shared\Transfer\UnzerApiResponseTransfer
+     */
+    public function convertUnzerApiGuzzleResponseToUnzerApiResponseTransfer(
+        UnzerApiResponseInterface $response,
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php]:
```suggestion
        UnzerApiResponseInterface $unzerApiResponse,
```

File: src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\UnzerApiResponseTransfer $responseTransfer
+     * @param array|null $responseData
+     *
+     * @return \Generated\Shared\Transfer\UnzerApiResponseTransfer
+     */
+    protected function updateResponseTransferWithError(
+        UnzerApiResponseTransfer $responseTransfer,
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php]:
```suggestion
        UnzerApiResponseTransfer $unzerApiResponseTransfer,
```

File: src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+            ->setIsSuccess(false)
+            ->setErrorResponse($errorTransfer);
+    }
+
+    /**
+     * @param bool $isSuccess
+     *
+     * @return \Generated\Shared\Transfer\UnzerApiResponseTransfer
+     */
+    protected function createResponseTransfer(bool $isSuccess): UnzerApiResponseTransfer
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php]:
```suggestion
    protected function createUnzerApiResponseTransfer(bool $isSuccess): UnzerApiResponseTransfer
```

File: src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Mapper/UnzerApiResponseMapperInterface.php [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+{
+    /**
+     * @param array $responseData
+     * @param \Generated\Shared\Transfer\UnzerApiResponseTransfer $responseTransfer
+     *
+     * @return \Generated\Shared\Transfer\UnzerApiResponseTransfer
+     */
+    public function mapResponseDataToUnzerApiResponseTransfer(
+        array $responseData,
+        UnzerApiResponseTransfer $responseTransfer
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Mapper/UnzerApiResponseMapperInterface.php]:
```suggestion
        UnzerApiResponseTransfer $unzerApiResponseTransfer
```

File: src/SprykerEco/Zed/UnzerApi/Dependency/External/UnzerApiToGuzzleAdapter.php [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+ */
+
+namespace SprykerEco\Zed\UnzerApi\Dependency\External;
+
+use GuzzleHttp\Client;
+use GuzzleHttp\Exception\RequestException;
+use GuzzleHttp\RequestOptions;
+use Psr\Http\Message\ResponseInterface;
+
+class UnzerApiToGuzzleAdapter implements UnzerApiToHttpClientInterface
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Dependency/External/UnzerApiToGuzzleAdapter.php]:
```suggestion
class UnzerApiToGuzzleHttpClientAdapter implements UnzerApiToHttpClientInterface
```

File: src/SprykerEco/Zed/UnzerApi/Dependency/External/UnzerApiToHttpClientException.php [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Zed\UnzerApi\Dependency\External;
+
+use Exception;
+use Throwable;
+
+class UnzerApiToHttpClientException extends Exception
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Dependency/External/UnzerApiToHttpClientException.php]: I think there's no need to have an Exception in the `Dependency` namespace.

File: src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+     * @var \SprykerEco\Zed\UnzerApi\Business\Api\Response\Mapper\UnzerApiResponseMapperInterface
+     */
+    protected $unzerApiResponseMapper;
+
+    /**
+     * @param \Spryker\Service\UtilEncoding\UtilEncodingServiceInterface $encodingService
+     * @param \SprykerEco\Zed\UnzerApi\Business\Api\Response\Mapper\UnzerApiResponseMapperInterface $responseMapper
+     */
+    public function __construct(
+        UtilEncodingServiceInterface $encodingService,
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php]:
```suggestion
        UtilEncodingServiceInterface $utilEncodingService,
```

File: src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+        UnzerApiResponseMapperInterface $responseMapper
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php]:
```suggestion
        UnzerApiResponseMapperInterface $unzerApiResponseMapper
```

File: src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    /**
+     * @param UnzerApiToHttpResponseInterface $unzerApiToHttpResponse
+     * @param bool $isSuccess
+     *
+     * @return \Generated\Shared\Transfer\UnzerApiResponseTransfer
+     */
+    public function convertUnzerApiGuzzleResponseToUnzerApiResponseTransfer(
+        UnzerApiToHttpResponseInterface $unzerApiToHttpResponse,
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/Business/Api/Response/Converter/UnzerApiResponseConverter.php]:
```suggestion
        UnzerApiToHttpResponseInterface $httpResponse,
```

File: src/SprykerEco/Zed/UnzerApi/UnzerApiDependencyProvider.php [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+{
+    /**
+     * @var string
+     */
+    public const SERVICE_UTIL_ENCODING = 'SERVICE_UTIL_ENCODING';
+
+    /**
+     * @var string
+     */
+    public const UNZER_API_HTTP_CLIENT = 'UNZER_API_HTTP_CLIENT';
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/UnzerApiDependencyProvider.php]:
```suggestion
    public const HTTP_CLIENT = 'HTTP_CLIENT';
```

File: src/SprykerEco/Zed/UnzerApi/UnzerApiDependencyProvider.php [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+
+        return $container;
+    }
+
+    /**
+     * @param \Spryker\Zed\Kernel\Container $container
+     *
+     * @return \Spryker\Zed\Kernel\Container
+     */
+    protected function addUnzerApiHttpClient(Container $container): Container
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/UnzerApi/UnzerApiDependencyProvider.php]:
```suggestion
    protected function addHttpClient(Container $container): Container
```

**Review by AsonUnique — APPROVED** [review]

**Pull request merged**